Added support for static plugins (#2943)
expo/expo-cli: packages/config-plugins/README.md

The Expo config is a powerful tool for generating native app code from a unified JavaScript interface. Most basic functionality can be controlled by using the the [static Expo config](https://docs.expo.io/versions/latest/config/app/), but some features require manipulation of the native project files. To support complex behavior we've created config plugins, and mods (short for modifiers).

> > 💡 **Hands-on Learners**: Use [this sandbox][sandbox] to play with the core functionality of Expo config. For more complex tests, use a local Expo project, with `expo eject --no-install` to apply changes.

Reviewer: ```suggestion
> 💡 **Hands-on Learners**: Use [this sandbox][sandbox] to play with the core functionality of Expo config plugins. For more complex tests, use a local Expo project, with `expo eject --no-install` to apply changes.
```

- [Usage](#usage)
- [What are plugins](#what-are-plugins)
- [Creating a plugin](#creating-a-plugin)
  - [Importing plugins](#importing-plugins)
  - [Chaining plugins](#chaining-plugins)
- [What are mods](#what-are-mods)
- [How mods works](#how-mods-works)
  - [Default mods](#default-mods)
  - [Mod plugins](#mod-plugins)
- [Creating a mod](#creating-a-mod)
  - [Experimental functionality](#experimental-functionality)
- [Plugin module resolution](#plugin-module-resolution)
  - [Project file](#project-file)
  - [app.plugin.js](#apppluginjs)
  - [Node module default file](#node-module-default-file)
  - [Project folder](#project-folder)
  - [Module internals](#module-internals)
  - [Raw functions](#raw-functions)
- [Why app.plugin.js for plugins](#why-apppluginjs-for-plugins)

**Quick facts**

- Plugins are functions that can change values on your Expo config.
- Plugins are mostly meant to be used with [`expo eject`][cli-eject] or `eas build` commands.
- We recommend you use plugins with `app.config.json` or `app.config.js` instead of `app.json` (no top-level `expo` object is required).

> - `mods` are async functions that modify native files.

Reviewer: ```suggestion
- `mods` are async functions that modify native project files, such as source code or configuration (plist, xml) files.
```

> - Changes performed with `mods` will require a native rebuild.

Reviewer: ```suggestion
- Changes performed with `mods` will require rebuilding the affected native projects.
```

#### Chaining Plugins

> Once you add a few plugins, your `app.plugin.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.

Author: ```suggestion
Once you add a few plugins, your `app.config.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.
```
```suggestion
Once you add a few plugins, your `app.plugin.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.
```

> Once you add a few plugins, your `app.plugin.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.

Author: ```suggestion
Once you add a few plugins, your `app.config.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.
```

export default withMySDK(config, { apiKey: 'X-XXX-XXX' });
```

### Importing plugins

You may want to create a plugin in a different file, here's how:

- The root file can be any JS file or a file named `app.plugin.js` in the [root of a Node module](#root-app.plugin.js).

> - The file should export a `ConfigPlugin` function.

Reviewer: ```suggestion
- The file should export a function that satisfies the `ConfigPlugin` type.
```

Reviewer: should we add the type definition to this readme or link to it?

Be careful using `withDangerousModifier` as it is subject to change in the future.
The order with which it gets executed is not reliable either.
Currently dangerous mods run first before all other modifiers, this is because we use dangerous mods internally for large file system refactoring like when the package name changes.

## Static Plugins

You don't need to migrate away from your existing `app.json` or `app.config.json` to use config plugins (although it is recommended). Plugins can be resolved automatically using the JSON `plugins` array:

```js
{
  name: 'My App',
  plugins: [
    // A config plugin!
    'expo-splash-screen'
  ]
}
```

Static plugins can be added in two different formats:

```js
{
  plugins: [
    // Long form with properties
    {
      resolve: 'my-plugin',
      props: {
        /* Passed as the second parameter to the config plugin */
      },

>     },

Author: An alternative API that could be more inline with config plugins is:
```js
{
  plugins: [
    ['my-plugin', { /* props */}],
    './my-other-plugin',
  ]
}
```

This could be extended to work like `withPlugins`

```js

const withSomething = (config, props) => config;

{
  plugins: [
    [withSomething, {}],
    withSomething,
    'expo-splash-screen'
  ]
}
```

Be careful using `withDangerousModifier` as it is subject to change in the future.
The order with which it gets executed is not reliable either.
Currently dangerous mods run first before all other modifiers, this is because we use dangerous mods internally for large file system refactoring like when the package name changes.

## Static Plugins

You don't need to migrate away from your existing `app.json` or `app.config.json` to use config plugins (although it is recommended). Plugins can be resolved automatically using the JSON `plugins` array:

```js
{
  name: 'My App',
  plugins: [
    // A config plugin!
    'expo-splash-screen'
  ]
}
```

Static plugins can be added in two different formats:

```js
{
  plugins: [
    // Long form with properties
    ['my-plugin', { /* Passed as the second parameter to the config plugin */ }]
    // Short hand -- gets normalized to ['my-plugin', undefined]
    'my-plugin',
  ];
}
```

> > 💡 Everything in the Expo config must be able to be converted to JSON (with the exception of the `mods` field). So no async functions!

Reviewer: ```suggestion
> 💡 Everything in the Expo config must be able to be converted to JSON (with the exception of the `mods` field). So no async functions outside of `mods` in your config plugins!
```

### Module Resolution

Static plugins can be resolved in a few different ways. Here are the different patterns for strings you could pass in the `plugins` array (the first property).

> Any resolution pattern that isn't specified below is unexpected behavior, and subject to breaking changes.

#### Project file

`'./my-config-plugin.js'`

```
╭── app.config.js ➡️ Expo Config
╰── my-config-plugin.js ➡️ ✅ `module.exports = (config) => config`
```

#### Node resolve

`'expo-splash-screen'`

```
╭── app.config.js ➡️ Expo Config
╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).
    ├── package.json ➡️ The `main` file points to `build/index.js`
    ╰── build/index.js ➡️  ✅ Node resolves to this module.
```

> Sometimes you want your package to export React components and also support a plugin, to support this, multiple entry points are used. If a `app.plugin.js` file is present in the Node module's root folder, it'll be used instead of the package's main file.

Reviewer: maybe we should move this above the `build/index.js` example, because it takes precedence

```
╭── app.config.js ➡️ Expo Config
╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).

>     ├── package.json ➡️ The `main` file will be used if `index.expo-plugin.js` doesn't exist.

Reviewer: ```suggestion
    ├── package.json ➡️ The `main` file will be used if `app.plugin.js` doesn't exist.
```

    ├── app.plugin.js ➡️ ✅ `module.exports = (config) => config` -- must export a function.

>     ╰── build/index.js ➡️ ❌ Ignored because `index.expo-plugin.js` exists. This could be used with `expo-splash-screen/build/index.js`

Reviewer: ```suggestion
    ╰── build/index.js ➡️ ❌ Ignored because `app.plugin.js` exists. This could be used with `expo-splash-screen/build/index.js`
```

```

#### Project folder

`'./my-config-plugin'`

> This is different to how Node module's work because `app.plugin.js` won't be resolved by default in a folder. You'll have to manually specify the file to use it.

Reviewer: ```suggestion
This is different to how Node modules work because `app.plugin.js` won't be resolved by default in a directory. You'll have to manually specify `./my-config-plugin/app.plugin.js` to use it, otherwise `index.js` in the directory will be used.
```

```
╭── app.config.js ➡️ Expo Config
╰── my-config-plugin/ ➡️ Folder containing plugin code

>     ╰── index.js ➡️ ✅ By default, node resolves a folder's index.js file as the main file.

Reviewer: ```suggestion
    ╰── index.js ➡️ ✅ By default, Node resolves a folder's index.js file as the main file.
```

for consistency with casing above

```

#### Module internals

> If a file inside a Node module is specified, then the module's root `app.plugin.js` resolution will be skipped. This is referred to as "reaching inside a package" and is bad form. We support this to make testing, and plugin authoring easier.

Reviewer: ```suggestion
If a file inside a Node module is specified, then the module's root `app.plugin.js` resolution will be skipped. This is referred to as "reaching inside a package" and is bad form. We support this to make testing, and plugin authoring easier, but we don't expect library authors to expose their plugins like this as a public API.
```

- `'expo-splash-screen/build/index.js'`
- `'expo-splash-screen/build'`

```
╭── app.config.js ➡️ Expo Config

> ╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).

Reviewer: ```suggestion
╰── node_modules/expo-splash-screen/ ➡️ Module installed from npm (works with Yarn workspaces as well).
```

npm spells their company name in all lower case

>     ├── package.json ➡️ The `main` file will be used if `index.expo-plugin.js` doesn't exist.

Reviewer: ```suggestion
    ├── package.json ➡️ The `main` file will be used if `app.plugin.js` doesn't exist.
```

    ├── app.plugin.js ➡️ ❌ Ignored because the reference reaches into the package internals.
    ╰── build/index.js ➡️ ✅ `module.exports = (config) => config`
```

#### Functions

You can also just pass in a config plugin.

```js
const withCustom = (config, props) => config;

const config = {
  plugins: [
    [
      withCustom,
      {
        /* props */
      },
    ],
    // Without props
    withCustom,
  ],
};
```

One caveat to using functions instead of strings is that serialization will replace the function with the function's name. This keeps manifests working as expected.

```json
{
  "plugins": [["withCustom", {}], "withCustom"]
}
```

### Why app.plugin.js for plugins

Config resolution searches for a `app.plugin.js` first when a Node module name is provided.

> This is because node environments are often different to iOS, Android, or web JS environments and therefore require different transpilation presets (ex: `module.exports` instead of `import/export`).

Reviewer: ```suggestion
This is because Node environments are often different to iOS, Android, or web JS environments and therefore require different transpilation presets (ex: `module.exports` instead of `import/export`).
```

Because of this reasoning, the root of a Node module is searched instead of right next to the `index.js`. Imagine you had a TypeScript Node module where the transpiled main file was located at `build/index.js`, if Expo config plugin resolution searched for `build/app.plugin.js` you'd lose the ability to transpile the file differently.

### Creating static plugins

> - The root file can be any JS file or the root `app.plugin.js` in a Node module.

Reviewer: not sure what the right wording here is but i think this is the first time mentioning "root file" so we might want to reword this a bit. maybe like "The entry point to your plugin can be ..."

- The file should export a `ConfigPlugin` function.

> - Plugins should be transpiled for Node environments ahead of time!

Reviewer: maybe we should recommend supporting all LTS releases and the current release of node?

> - Plugins should be transpiled for Node environments ahead of time!

Author: ```suggestion
- Plugins should be transpiled for Node environments ahead of time! They should support the versions of Node that [Expo supports](https://docs.expo.io/get-started/installation/#requirements) (LTS).
```

    [withBar, 'input 1'],
    [withFoo, 'input 2'],
    [withDelta, 'input 3'],
  ],
};
```

## What are mods

An async function which accepts a config and a data object, then manipulates and returns both as an object.

Modifiers (mods for short) are added to the `mods` object of the Expo config. The `mods` object is different to the rest of the Expo config because it doesn't get serialized after the initial reading, this means you can use it to perform actions _during_ code generation. If possible, you should attempt to use basic plugins instead of mods as they're simpler to work with.

- `mods` are omitted from the manifest and **cannot** be accessed via `Updates.manifest`. mods exist for the sole purpose of modifying native files during code generation!
- `mods` can be used to read and write files safely during the `expo eject` command. This is how Expo CLI modifies the Info.plist, entitlements, xcproj, etc...
- `mods` are platform specific and should always be added to a platform specific object:

`app.config.js`

```js
module.exports = {
  name: 'my-app',
  mods: {
    ios: {
      /* iOS mods... */
    },
    android: {
      /* Android mods... */
    },
  },
};
```

> ## How mods works

Reviewer: ```suggestion
## How mods work
```
